TelepresenceBot/mobile/src/main/java/com/novoda/tpbot/bot/SocketIOTpService.java

```diff
@@ -0,0 +1,91 @@
+package com.novoda.tpbot.bot;
+
+import android.os.Handler;
+import android.os.Looper;
+
+import com.novoda.tpbot.Direction;
+import com.novoda.tpbot.Result;
+import com.novoda.tpbot.support.Event;
+import com.novoda.tpbot.support.MalformedServerAddressException;
+import com.novoda.tpbot.support.Observable;
+import com.novoda.tpbot.support.SocketConnectionObservable;
+
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+import io.socket.client.IO;
+import io.socket.client.Socket;
+import io.socket.emitter.Emitter;
+
+class SocketIOTpService implements BotTpService {
+
+    private Socket socket;
+    private final Handler handler;
+
+    static SocketIOTpService getInstance() {
+        return LazySingleton.INSTANCE;
+    }
+
+    private SocketIOTpService() {
+        this.handler = new Handler(Looper.getMainLooper());
+    }
+
+    @Override
+    public Observable<Result> connectTo(String serverAddress) {
+        try {
+            URL url = new URL(serverAddress);
+            socket = IO.socket(url.toExternalForm());
+        } catch (MalformedURLException | URISyntaxException exception) {
+            MalformedServerAddressException exceptionWithUserFacingMessage = new MalformedServerAddressException(exception);
+            return Observable.just(Result.from(exceptionWithUserFacingMessage));
+        }
+        return new SocketConnectionObservable(socket, handler, Event.JOIN_BOT);
+    }
+
+    @Override
+    public Observable<Direction> listen() {
+        return new DirectionObservable();
+    }
+
+    private class DirectionObservable extends Observable<Direction> {
+
+        @Override
+        public Observable<Direction> start() {
+            if (!socket.connected()) {
```

Comment:
It has automatic retry built in. I'm handling the case when we get disconnected with a `Result.error` which will make the user connect to the server again.